solana-program/js/test/tests.ts

```diff
@@ -0,0 +1,232 @@
+    throw new Error(
+      `Failed to read program keypair at '${PROGRAM_KEYPAIR_PATH}' due to error: ${errMsg}.`,
+    );
+  }
+
+  // Check if the program has been deployed
+  const programInfo = await connection.getAccountInfo(programId);
+  if (programInfo === null) {
+    if (fs.existsSync(PROGRAM_SO_PATH)) {
+      throw new Error(
+        'Program needs to be deployed with `npm run deploy_program`',
+      );
+    } else {
+      throw new Error('Program needs to be built and deployed');
+    }
+  } else if (!programInfo.executable) {
+    throw new Error(`Program is not executable`);
+  }
+  console.log(`Using program ${programId.toBase58()}`);
+}
+
+/**
+ * Establish a new tokenToDump owned by the payer account, 
+ * to be dumped into trashpile
+ */
+let tokenToDumpMintAddress: PublicKey;
+let tokenToDumpSrcAddress: PublicKey;
+let tokenToDumpDest: Account;
+export async function establishTokenToDump(): Promise<void> {
+  tokenToDumpMintAddress = await createMint(
+    connection,
+    payer,
+    payer.publicKey,
+    null,
+    9,
+  );
+  console.log(
+    'Using tokenToDump with mint ',
+    tokenToDumpMintAddress.toBase58()
+  );
+
+  tokenToDumpSrcAddress = await createAccount(
+    connection,
+    payer,
+    tokenToDumpMintAddress,
+    payer.publicKey
+  );
+  console.log('Using ', tokenToDumpSrcAddress.toBase58(), 'to originate tokenToDump');
+  await mintTo(
+    connection,
+    payer,
+    tokenToDumpMintAddress,
+    tokenToDumpSrcAddress,
+    payer,
+    1
+  );
+  console.log('Minted 1 tokenToDump to ', tokenToDumpSrcAddress.toBase58());
+
+  tokenToDumpDest = await getOrCreateAssociatedTokenAccount(
+    connection,
+    payer,
+    tokenToDumpMintAddress,
+    programId,
+  )
+  console.log('Using PDA ', tokenToDumpDest.address.toBase58(), 'to receive tokenToDump');
+}
+
+/**
+ * Create trash token mint where authority is PDA of trashpile program
+ */
+let trashTokenMintAuthorityAddress: PublicKey;
+let trashTokenMintAuthorityBumpSeed: number;
+let trashTokenMintAddress: PublicKey;
+let trashTokenDest: Account;
+export async function establishTrashToken(): Promise<void> {
+  [trashTokenMintAuthorityAddress, trashTokenMintAuthorityBumpSeed] = await PublicKey.findProgramAddress(
+    [Buffer.from("$TRASH")],
+    programId
+  );
+
+  trashTokenMintAddress = await getOrCreateTrashTokenMintAddress(
+    connection,
+    payer,
+    trashTokenMintAuthorityAddress
+  );
+  console.log(
+    'Using trash token with mint ',
+    trashTokenMintAddress.toBase58()
+  );
+  trashTokenDest = await getOrCreateAssociatedTokenAccount(
+    connection,
+    payer,
+    trashTokenMintAddress,
+    payer.publicKey,
+  )
+  console.log('Using ', trashTokenDest.address.toBase58(), 'to receive trash token')
+}
+
+/**
+ * Dump tokenToDump, receive trash token in exchange
+ */
+export async function dumpTokenToDump(): Promise<TransactionSignature> {
+
+  console.log('Forming dump instruction...')
```

Comment:
> 1 - should this function take in a specific NFT token as an input? Otherwise, how do you know which NFT to dump?

In the test here, we're just generating a random new NFT via `establishTokenToDump` and storing that as a global, hence no arguments to `dumpTokenToDump`. The actual solana program, though, very much does take specific token inputs: these are the `token_to_dump_...` keys below.

Note that you'll also need to pass `trash_token_...` stuff and `token_program` (the spl `TOKEN_PROGRAM_ID`), and that `tokenToDumpDest` (and `trashTokenDest`) must be created by the dumper at runtime if not preexisting (see `getOrCreateAssociatedTokenAccount` calls above). It's easier/cheaper to do all this stuff off-chain.

> 2 - How about the 'multiple NFTs' case (I select many NFTs to dump)?

Multi-token dumps should be constructed in JS by adding multiple dump instructions (one per token) to a single signed instruction. Currently this will return 1 $TRASH per instruction.

Reply:
> Multi-token dumps should be constructed in JS by adding multiple dump instructions (one per token) to a single signed instruction (tx.add(ix); --> sendAndConfirmTransaction). Currently this will return 1 $TRASH per instruction.

OK. I want to avoid the case where a user needs to 'confirm' multiple transactions.

Will accept the PR.